rfcs/20201221-tfmot-compression-api.md

```diff
@@ -0,0 +1,441 @@
+
+## User Benefit
+
+In this design, we'd like to reduce the common engineering cost for the compression algorithm developers.
+
+* Write unit test model coverage test, and benchmark. Provide the comparisons of compression algorithms.
+* Deployment compressed model. (TF serving, TFLite, and tf.js)
+* Support TF 2.0 Keras features compatibility. (e.g. distributed training.)
+
+## Design Proposal
+
+We propose the compression algorithm API which helps algorithm developers create model developer APIs for their own compression algorithm.
+Our API also provides guidelines for testing and benchmark. For now, we only have guidelines to apply a compression algorithm for simple MNIST vision cases. We'd like to provide an example for tensorflow [official models](https://github.com/tensorflow/models/tree/master/official) in the future.
+
+### Tutorials and Examples
+We provide the tutorial for [SVD](https://en.wikipedia.org/wiki/Singular_value_decomposition) compression algorithm that shows how we implement the SVD algorithm using TFMOT compression API by colab. This tutorial includes:
+
+#### Algorithm developer side
+The algorithm developer implementing the SVD algorithm uses the `WeightCompressor` class. It also includes a custom model developer API for the SVD algorithm.
+
+```python
+class SVD(algorithm.WeightCompressor):
+  """SVD compression module config."""
+
+  def __init__(self, rank):
+    self.rank = rank
+
+  def init_training_weights(
+      self, pretrained_weight: tf.Tensor):
+    """Init function from pre-trained model case."""
+    # Dense Layer
+    if len(pretrained_weight.shape) == 2:
+      u, sv = tf_svd_factorization_2d(pretrained_weight, self.rank)
+    else:
+      raise NotImplementedError('Only for dimension=2 is supported.')
+
+    self.add_training_weight(
+        name='u',
+        shape=u.shape,
+        dtype=u.dtype,
+        initializer=tf.keras.initializers.Constant(u))
+    self.add_training_weight(
+        name='sv',
+        shape=sv.shape,
+        dtype=sv.dtype,
+        initializer=tf.keras.initializers.Constant(sv))
+
+  def project_training_weights(self, u: tf.Tensor, sv: tf.Tensor) -> tf.Tensor:
+    return tf.matmul(u, sv)
+
+  def get_compressible_weights(
+      self, original_layer: tf.keras.layers.Layer) -> List[tf.Variable]:
+    if isinstance(original_layer, tf.keras.layers.Dense):
+      input_dim = original_layer.kernel.shape[0]
+      output_dim = original_layer.kernel.shape[1]
+      if input_dim * output_dim > (input_dim + output_dim) * self.rank:
+        return [original_layer.kernel]
+    return []
+
+  def optimize(self, to_optimize: tf.keras.Model) -> tf.keras.Model:
```

Comment:
I feel most people would interpret `optimize` to mean "adjust the weights to give lower loss" (that's what keras `Optimizer`s do after all). Using the term to mean something completely different here seems uneccessarily confusing to me. `compressed_model` maybe?

Also, is this supposed to be part of the `WeightCompressor` interface as well?

Reply:
Agree that `compress` is more explicit; `optimize` actually conflicts with the optimizer method with the same name.

Reply:
I'm not sure how much people are bothered, but `Optimizer.optimize` mutates the argument, where as `WeightCompressor.compress` woiuld leave the input unchanged and return a compressed version. I've seen past-tense used elsewhere (`compressed`) to disambiguate.

Reply:
`get_compressed_model` would clarify this but is also more verbose. I think this should not be a point of confusion in general. The closest thing we have in the API is `clone_model` which also returns a new model. So `compress` is likely fine.

Reply:
I think `optimize` is not a part of `WeightCompressor` because it's a model developer user interface that each compression algorithm can have a different interface. (e.g. For the SVD case we only have one step (compress_model) API, but some compression algorithms like Pruning require an extra step to make tensor dense to sparse after fine-tuning.)

`compress_model` seems clear for now. I changed the method name. Please keep discussing if you have any idea about this. Thanks!